### PR TITLE
PWA: migrate document info to ampdoc-level singleton

### PR DESCRIPTION
**Pull request by dvoytenko**
Partial for #3742.

### CODE REVIEW

File: extensions/amp-app-banner/0.1/amp-app-banner.js [hunk 1]
```diff
@@ -381,7 +381,7 @@ export class AmpAndroidAppBanner extends AbstractAppBanner {
 
   /** @private */
   getAndroidIntentForUrl_(appId) {
-    const canonicalUrl = documentInfoFor(this.win).canonicalUrl;
+    const canonicalUrl = documentInfoForDoc(this.element).canonicalUrl;
```
Comment (jridgewell) [inline extensions/amp-app-banner/0.1/amp-app-banner.js]: It's kinda weird that we call this `documentInfoForDoc` but pass in an `Element`.

File: src/document-info.js [hunk 2]
```diff
@@ -14,31 +14,51 @@
...
-              'AMP files are required to have a <link rel=canonical> tag.')
-              .href).href,
-      pageViewId: getPageViewId(win),
-      sourceUrl: getSourceUrl(win.location.href),
-    };
+export function documentInfoForDoc(nodeOrDoc) {
+  return getServiceForDoc(nodeOrDoc, 'documentInfo', ampdoc => {
+    const url = ampdoc.getUrl();
+    const sourceUrl = getSourceUrl(url);
+    let canonicalUrl = ampdoc.getRootNode() && ampdoc.getRootNode().AMP &&
```
Comment (jridgewell) [inline src/document-info.js]: Might be a bit cleaner to store `var rootNode = ampdoc.getRootNode()`

**Comment by jridgewell** [comment]
LGTM.

File: extensions/amp-app-banner/0.1/amp-app-banner.js (hunk 1 above)
Comment (dvoytenko) [inline extensions/amp-app-banner/0.1/amp-app-banner.js]: Yeah. We went back and force on it. Basically, it starts with the context element which points itself directly at a doc. E.g. `documentInfoForDoc(ampdocService.findAmpDocForElement(this.element))`. Constantly doing this was just two much. So it's hidden behind that API.

File: src/document-info.js (hunk 2 above)
Comment (dvoytenko) [inline src/document-info.js]: Done

File: src/service/url-replacements-impl.js [hunk 3]
```diff
@@ -407,6 +407,15 @@ export class UrlReplacements {
   }
 
   /**
+   * Resolves the value via document info.
+   * @param {function(!DocumentInfoDef):*} getter
+   * @return {*}
```
Comment (erwinmombay) [inline src/service/url-replacements-impl.js]: why `*` wouldnt this be `DocumentInfoDef`?

Comment (erwinmombay) [inline src/service/url-replacements-impl.js]: nvm didnt see the getter

**Comment by erwinmombay** [comment]
LGTM

File: src/service/url-replacements-impl.js (hunk 3 above)
Comment (jridgewell) [inline src/service/url-replacements-impl.js]: `@template T` might be better here.

Comment (dvoytenko) [inline src/service/url-replacements-impl.js]: Yes. `@template` is better. Replaced it here and elsewhere in this module.

**Pull request merged**